File: cpp/src/app/options/OptionsGroupRaytracingMethod.cpp
#include <cstring>

#include "common/commandLineOptions/OptionParser.h"
#include "common/commandLineOptions/TypedOption.h"
#include "app/options/OptionsGroupRaytracingMethod.h"

char *OptionsGroupRaytracingMethod::raytracingMethodsString = nullptr;
char *OptionsGroupRaytracingMethod::rayTracerName = nullptr;

void
OptionsGroupRaytracingMethod::mainRayTracingOption(char *&name) {
    strcpy(rayTracerName, name);
}

void
OptionsGroupRaytracingMethod::rayTracingParseOptions(
        int *argc,
        char **argv,
        char raytracingMethodsStringOut[],
        char *rayTracerNameOut)
{
    char *raytracingMethodName = nullptr;
    TypedOption<char *> raytracingMethodOpt = {"-raytracing-method", &raytracingMethodName, 1, OptionsGroupRaytracingMethod::mainRayTracingOption, nullptr};
    OptionBase raytracingOptions[] = {
        REGISTER_OPTION(char *, raytracingMethodOpt, 4)
    };

    OptionsGroupRaytracingMethod::rayTracerName = rayTracerNameOut;
    OptionsGroupRaytracingMethod::raytracingMethodsString = raytracingMethodsStringOut;
    OptionGroup raytracingGroups[] = {
        OptionGroup("raytracing", raytracingOptions, 1)
    };
    OptionParser<OptionBase>::parse(argc, argv, raytracingGroups, 1);
}

File: typescript/src/main/vsdk/toolkit/app/options/OptionsGroupRaytracingMethod.ts
import { OptionBase } from "../../common/commandLineOptions/OptionBase";
import { OptionGroup } from "../../common/commandLineOptions/OptionGroup";
import { OptionParser } from "../../common/commandLineOptions/OptionParser";
import { TypedOption } from "../../common/commandLineOptions/TypedOption";

export class OptionsGroupRaytracingMethod {
  private static raytracingMethodsString: string | null = null;
  private static rayTracerName: string[] | null = null;

  private constructor() {
  }

  private static mainRayTracingOption(name: TypedOption.MutableValue<string | null>): void {
    if (OptionsGroupRaytracingMethod.rayTracerName !== null && OptionsGroupRaytracingMethod.rayTracerName.length > 0) {
      OptionsGroupRaytracingMethod.rayTracerName[0] = name.value ?? "";
    }
  }

  public static rayTracingParseOptions(
    argc: number[],
    argv: string[],
    raytracingMethodsStringOut: string[],
    rayTracerNameOut: string[]
  ): void {
    const raytracingMethodName = TypedOption.valueRef<string | null>(null);
    const raytracingMethodOpt = new TypedOption<string | null>(
      "-raytracing-method",
      raytracingMethodName,
      1,
      OptionsGroupRaytracingMethod.mainRayTracingOption,
      null
    );
    const raytracingOptions: OptionBase[] = [
      TypedOption.REGISTER_OPTION(raytracingMethodOpt, 4),
    ];

    OptionsGroupRaytracingMethod.rayTracerName = rayTracerNameOut;
    OptionsGroupRaytracingMethod.raytracingMethodsString =
      raytracingMethodsStringOut !== null && raytracingMethodsStringOut.length > 0
        ? raytracingMethodsStringOut[0]
        : null;
    const raytracingGroups = [
      new OptionGroup("raytracing", raytracingOptions, 1),
    ];
    OptionParser.parse(argc, argv, raytracingGroups, 1);
  }
}
